drivers/gpio/gpio_bflb_bl60x_70x.c

```diff
@@ -0,0 +1,386 @@
+		tmp_a |= BIT(pin);
+		if (flags & GPIO_OUTPUT_INIT_HIGH) {
+			tmp = sys_read32(cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+			tmp = tmp | pin;
+			sys_write32(tmp, cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+		}
+		if (flags & GPIO_OUTPUT_INIT_LOW) {
+			tmp = sys_read32(cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+			tmp = tmp & ~pin;
+			sys_write32(tmp, cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+		}
+	} else {
+		tmp_a &= ~BIT(pin);
+	}
+
+
+	sys_write32(tmp_a, cfg->base_reg + GLB_GPIO_CFGCTL34_OFFSET);
+
+	if ((flags & GPIO_PULL_UP) != 0) {
+		tmp_b |= BIT(is_odd * GPIO_BFLB_2ND_GPIO_POS + GLB_REG_GPIO_0_PU_POS);
+		tmp_b &= ~BIT(is_odd * GPIO_BFLB_2ND_GPIO_POS + GLB_REG_GPIO_0_PD_POS);
+	} else if ((flags & GPIO_PULL_DOWN) != 0) {
+		tmp_b |= BIT(is_odd * GPIO_BFLB_2ND_GPIO_POS + GLB_REG_GPIO_0_PD_POS);
+		tmp_b &= ~BIT(is_odd * GPIO_BFLB_2ND_GPIO_POS + GLB_REG_GPIO_0_PU_POS);
+	} else {
+		tmp_b &= ~BIT(is_odd * GPIO_BFLB_2ND_GPIO_POS + GLB_REG_GPIO_0_PU_POS);
+		tmp_b &= ~BIT(is_odd * GPIO_BFLB_2ND_GPIO_POS + GLB_REG_GPIO_0_PD_POS);
+	}
+
+	/* GPIO mode */
+#ifdef CONFIG_SOC_SERIES_BL70X
+	/* but function goes in the right place */
+	if (pin >= GPIO_BFLB_BL70X_PSRAM_START && pin <= GPIO_BFLB_BL70X_PSRAM_END) {
+		tmp = sys_read32(cfg->base_reg + GLB_GPIO_CFGCTL0_OFFSET
+			+ (pin / GPIO_BFLB_PIN_PER_WORD * GPIO_BFLB_WORDSIZE));
+		tmp &= ~(GPIO_BFLB_FUNC_MSK
+			<< ((pin & 1) * GPIO_BFLB_2ND_GPIO_POS + GLB_REG_GPIO_0_FUNC_SEL_POS));
+		tmp |= (GPIO_BFLB_FUNCTION_GPIO
+			<< ((pin & 1) * GPIO_BFLB_2ND_GPIO_POS + GLB_REG_GPIO_0_FUNC_SEL_POS));
+		sys_write32(tmp, cfg->base_reg + GLB_GPIO_CFGCTL0_OFFSET
+			+ (pin / GPIO_BFLB_PIN_PER_WORD * GPIO_BFLB_WORDSIZE));
+	} else {
+		tmp_b &= ~(GPIO_BFLB_FUNC_MSK
+			<< (is_odd * GPIO_BFLB_2ND_GPIO_POS + GLB_REG_GPIO_0_FUNC_SEL_POS));
+		tmp_b |= (GPIO_BFLB_FUNCTION_GPIO
+			<< (is_odd * GPIO_BFLB_2ND_GPIO_POS + GLB_REG_GPIO_0_FUNC_SEL_POS));
+	}
+#else
+	tmp_b &= ~(GPIO_BFLB_FUNC_MSK
+		<< (is_odd * GPIO_BFLB_2ND_GPIO_POS + GLB_REG_GPIO_0_FUNC_SEL_POS));
+	tmp_b |= (GPIO_BFLB_FUNCTION_GPIO
+		<< (is_odd * GPIO_BFLB_2ND_GPIO_POS + GLB_REG_GPIO_0_FUNC_SEL_POS));
+#endif
+	/* enabled SMT in GPIO mode */
+	tmp_b |= BIT(is_odd * GPIO_BFLB_2ND_GPIO_POS + GLB_REG_GPIO_0_SMT_POS);
+	tmp_b |= (GPIO_BFLB_DRIVE_STRENGTH
+		<< (is_odd * GPIO_BFLB_2ND_GPIO_POS + GLB_REG_GPIO_0_DRV_POS));
+
+	sys_write32(tmp_b, cfg_address);
+
+	return 0;
+}
+
+int gpio_bflb_init(const struct device *dev)
+{
+	const struct gpio_bflb_config * const cfg = dev->config;
+
+	cfg->irq_config_func(dev);
+
+	return 0;
+}
+
+static void gpio_bflb_isr(const struct device *dev)
+{
+	const struct gpio_bflb_config * const cfg = dev->config;
+	struct gpio_bflb_data *data = dev->data;
+	uint32_t int_stat;
+
+	/* interrupt data is in format 1 bit = 1 pin */
+	int_stat = sys_read32(cfg->base_reg + GLB_GPIO_INT_STAT1_OFFSET);
+
+	gpio_fire_callbacks(&data->callbacks, dev, int_stat);
+	gpio_bflb_pin_interrupt_clear(dev, int_stat);
```

Comment:
If the callback does something that triggers a new interrupt, calling `gpio_flbl_pin_interrupt_clear()` afterward would discard the new interrupt? The SDK does it the other way around.

https://github.com/bouffalolab/bouffalo_sdk/blob/811e8c1f1cd62695cedd47367f0bf2c624707bc6/drivers/lhal/src/bflb_gpio.c#L810-L811

Or I missed something again.